ENH: support cut/qcut for datetime/timedelta (GH14714)

aileronajay opened the pull request:
- [x] closes #14714
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master | flake8 --diff``
 - [x] whatsnew entry

Commit:
initial iteration for pd cut enhance to support time date type

aileronajay commented:
https://github.com/pandas-dev/pandas/issues/14714

aileronajay commented:
@jreback created this PR to verify if i am heading in the right direction with this change

jorisvandenbossche commented:
@aileronajay Roughly this looks in the right direction, yes. But, we will also want this to work with datetime and not only timedelta values. So you will have to pass the dtype somehow to distuinguish between both when making the labels.
Another problem will be that you now assume this to be a series (you access the `.values`), while this also works for arrays. And that you convert a series to an array, and in this way loose the index information.

I would also encourage you to already add tests with some examples and the desired results, so you know what you are working towards.

jreback commented on pandas/tools/tile.py:
```diff
@@ -249,16 +259,24 @@ def _format_levels(bins, prec, right=True,
                 raise ValueError('precision too low')
 
-            formatted = '(%s, %s]' % (fa, fb)
+            if time_data:
```
look at ``pandas/core/algorithms.py/factorize`` for how we handle the differing dtypes. we want to replicate that type of logic (what you are doing is fine, but the code will be more smooth).

aileronajay replied:
@jreback i have moved this handling down to the formatter in the latest iteration of this code

jreback commented:
pls add some tests

aileronajay commented on pandas/tools/tile.py:
```diff
@@ -81,6 +82,13 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
+    # for handling the cut for datetime and timedelta objects
+    if needs_i8_conversion(x):
+        x = x.values.view('i8')
+        time_data = True
```
@jreback the x.dtype call errors out with this message when i run nosetests on test_tile.py AttributeError: 'list' object has no attribute 'dtype'. The formatter that i create in the suggested change here will be used in lieu of the formatter that is present in the downstream code, in method _format_levels  , is that correct?

jreback replied:
use ``np.asarray``

aileronajay commented:
@jorisvandenbossche ". And that you convert a series to an array, and in this way loose the index information.
"
After making call to the cut, will the output still be related to the initial index? I thought we split the  data into ranges by making a call to cut and the result does not have the same order/index as the original input that we pass to cut

jorisvandenbossche commented:
@aileronajay As you can see, yes, the return value of `cut` retains the index:

```
In [18]: s = pd.Series(range(5))

In [19]: s
Out[19]: 
0    0
1    1
2    2
3    3
4    4
dtype: int64

In [20]: pd.cut(s, bins=2)
Out[20]: 
0    (-0.004, 2]
1    (-0.004, 2]
2    (-0.004, 2]
3         (2, 4]
4         (2, 4]
dtype: category
Categories (2, object): [(-0.004, 2] < (2, 4]]
```

Have a look for `x_is_series` how this is handled currently (eg here: https://github.com/aileronajay/pandas/blob/1b2dd2428d6ed2371cac89e56d474941c7225653/pandas/tools/tile.py#L243)

Commit:
added datetime handling and passing object type

aileronajay commented:
@jorisvandenbossche @jreback  i have changed how i convert the time object to integers, so i get a series now instead of array , x = x.astype(np.int64), it is similar to what is done in pandas.tools.util to_numeric method. This is the output that i  get now

 s = pd.Series(pd.to_timedelta(np.random.randint(0,10000,size=10),unit='ns')).sort_values()
 pd.cut(s,6)
1    (0 days 00:00:00.000000, 0 days 00:00:00.000001]
9    (0 days 00:00:00.000000, 0 days 00:00:00.000001]
7    (0 days 00:00:00.000001, 0 days 00:00:00.000003]
4    (0 days 00:00:00.000004, 0 days 00:00:00.000006]
2    (0 days 00:00:00.000004, 0 days 00:00:00.000006]
0    (0 days 00:00:00.000004, 0 days 00:00:00.000006]
3    (0 days 00:00:00.000004, 0 days 00:00:00.000006]
5    (0 days 00:00:00.000006, 0 days 00:00:00.000007]
8    (0 days 00:00:00.000006, 0 days 00:00:00.000007]
6    (0 days 00:00:00.000007, 0 days 00:00:00.000008]

Commit:
modified the formatter to handle time objects

Commit:
added testcase for datetime cut

aileronajay commented:
@jreback i have moved the extra processing to the format method

Commit:
added changes to qcut code

sinhrks commented on pandas/tools/tile.py:
```diff
@@ -81,6 +83,17 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
+
+    dtype = None
+    if is_timedelta64_dtype(x):
+        x = x.astype(np.int64)
```
pls use ``view(np.int64)``

aileronajay replied:
@sinhrks incorporated this change

sinhrks commented on pandas/tools/tile.py:
```diff
@@ -81,6 +83,17 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
     """
     # NOTE: this binning code is changed a bit from histogram for var(x) == 0
+    # for handling the cut for datetime and timedelta objects
+
```
better to convert to array likes here because input can be a ``list``.  prepare a private function which can be used in ``cut`` and ``qcut``.

sinhrks replied:
Yes, we should preserve the index. So needs below logic before converting to array.

- https://github.com/aileronajay/pandas/blob/b21fa237537ff225fca955355000a7b2fc2a2bb4/pandas/tools/tile.py#L204

jorisvandenbossche replied:
Possibly it will be easier to move the `x_is_series` logic up (to here, so you can then just deal with arrays)

aileronajay replied:
@jorisvandenbossche @sinhrks the logic that preserves the index is present in _bins_to_cuts method. Whereas the error that we get due to date/timedelta data is due to code that is present in the cut method. So we need to convert into int64 in the cut method itself. So i have two questions. The first is that do i need to convert it to array, that i originally started out doing by making a call to  x.values.view('i8') ? if yes, do i need to move to index handling code to the cut method instead of _bins_to_cuts method?

sinhrks replied:
Yes, I agree preprocessing/postprocessing should be moved to ``cut/qcut`` from ``_bins_to_cuts``. what on my mind is something like below. 

Actually preprocessing / postprocessing can be wrapped as a private function to reduce code duplications. 

```
def cut(...):
    x_is_series = ...
    ... 
    x = np.asarray(x)    
    if is_datetime64_dtype(...):
        ....

    (core logic)

    if x_is_series:
        ....
    return...
```

aileronajay replied:
@sinhrks I have created two private methods, _preprocess_for_cut and _postprocess_for_cut to move the logic from _bins_to_cuts to cut and qcut methods

sinhrks replied:
Thx for the change. I **think** ``_preprocess_for_cut`` should be called before ``_coerce_to_type`` to handle ``list`` properly?

aileronajay replied:
i think i can merge _coerce_to_type code with _preprocess_for_cut and return an additional dtype parameter from the _preprocess_for_cut method. @jreback are there any concerns if i remove the _coerce_to_type code method and the processing that is being done to _preprocess_for_cut method?

aileronajay replied:
@sinhrks i have moved the call to _preprocess_for_cut  prior to coerce

sinhrks commented on pandas/tools/tile.py:
```diff
@@ -258,12 +283,16 @@ def _format_levels(bins, prec, right=True,
+def _format_label(x, precision=3, dtype=None):
     fmt_str = '%%.%dg' % precision
+
+    if dtype == np.datetime64:
```
pls use ``is_datetime64_dtype`` here also.

aileronajay replied:
@sinhrks, i can do that, that means that i would not have to pass the dtype as an argument, right? (I can check the dtype using is_datetime64_dtype call). Though it would mean that i make this call twice, once early on, when qcut or cut is called and later in _format_label, i mean if the call is not expensive, it shouldnt make a lot of difference

sinhrks replied:
Shouldn't expensive. Alternative option is to use temporary variable. Including other answers, following impl is on my mind.

- https://github.com/pandas-dev/pandas/blob/master/pandas/core/algorithms.py#L400

aileronajay replied:
@sinhrks made this change now

sinhrks commented on pandas/tools/tile.py:
```diff
@@ -81,6 +83,17 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
+        x = x.astype(np.int64)
+        dtype = np.timedelta64
+
+    if is_datetime64_dtype(x):
```
It is nice if we can support datetimetz.

aileronajay replied:
and will that done be in the same same manner as well? (by converting to int64?)

Commit:
code review comments

jorisvandenbossche commented on pandas/tools/tests/test_tile.py:
```diff
@@ -283,6 +284,16 @@ def test_single_bin(self):
+        self.assertEqual(result[1],
+                         '(2013-01-01 16:00:00, 2013-01-02 08:00:00]')
+        self.assertEqual(result[2],
+                         '(2013-01-02 08:00:00, 2013-01-03 00:00:00]')
```
Instead of checking each element, it is better to construct the series, and then use `assert_series_equal`:

```
exp = Series(['(2012-12-31 23:57:07.200000, 2013-01-01 16:00:00]', '(2013-01-01 16:00:00, 2013-01-02 08:00:00]', '(2013-01-02 08:00:00, 2013-01-03 00:00:00]')])
tm.assert_series_equal(result, exp)
```

aileronajay replied:
@jorisvandenbossche @jreback incorporated your review comments

jreback commented on pandas/tools/tile.py:
```diff
@@ -11,8 +11,10 @@
 from pandas.compat import zip
-
+from pandas.tseries.timedeltas import to_timedelta
+from pandas import to_datetime
```
import to_timedelta from pandas directly

aileronajay replied:
@jreback i have made this change

jreback commented on pandas/tools/tile.py:
```diff
@@ -81,6 +83,17 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
+        x = x.view(np.int64)
+        dtype = np.timedelta64
+
+    if is_datetime64_dtype(x):
```
should be elseif

aileronajay replied:
@jreback i have made this change

jreback commented on pandas/tools/tile.py:
```diff
@@ -258,12 +283,16 @@ def _format_levels(bins, prec, right=True,
     fmt_str = '%%.%dg' % precision
+
+    if is_datetime64_dtype(dtype):
+        return to_datetime(x, unit='ns')
```
i would like to see thur format_labels so to do all labels in 1 go

aileronajay replied:
@jreback can you please explain, i am not getting what you are trying to say

Commit:
code review comments

jreback commented on pandas/tools/tile.py:
```diff
@@ -11,8 +11,10 @@
 import pandas.core.nanops as nanops
 from pandas.compat import zip
-
+from pandas import to_timedelta
```
put on the same line

aileronajay replied:
all import from pandas are on the same line

jreback commented on pandas/tools/tile.py:
```diff
@@ -11,8 +11,10 @@
+from pandas import to_timedelta
+from pandas import to_datetime
 import numpy as np
+from pandas.types.common import (is_datetime64_dtype, is_timedelta64_dtype)
```
parens not needed

aileronajay replied:
parenthesis removed now

jreback commented on pandas/tools/tile.py:
```diff
@@ -81,6 +83,17 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
+
+    elif is_datetime64_dtype(x):
+        x = x.view(np.int64)
+        dtype = np.datetime64
```
make a function in this module like

```
def _coerce_to_type(x):
     dtype = None
     original = x
     x = np.asarray(x)
     if is_timedelta64_dtype(x):
....

    return original, x, dtype
```
Then call this (in the 2 places u r using it).

return the orignal data, coerced x, dtype

then also pass the original data (if you need to reconstruct something in side the functions)

aileronajay replied:
i have implemented this change now

Commit:
added private method for coercing data

aileronajay committed:
added private methods for pre and post processing

sinhrks commented on pandas/tools/tests/test_tile.py:
```diff
@@ -283,6 +284,16 @@ def test_single_bin(self):
+    def test_datetime_cut(self):
+        # GH 14714
+        data = to_datetime(Series(['2013-01-01', '2013-01-02', '2013-01-03']))
+        result, bins = cut(data, 3, retbins=True)
```
can u also test:

- ``list`` of ``np.datetime64``
- ``ndarray`` of ``datetime64`` dtype
- ``DatetimeIndex``

aileronajay committed:
moving preprocess prior to coerce to type

aileronajay committed:
rectified preprocess method array conversion missing

aileronajay committed:
added tests and modified preprocessing code

aileronajay committed:
added test for datetime index

aileronajay commented:
@sinhrks @jreback @jorisvandenbossche i have incorporated changes pertaining to all the review comments

jorisvandenbossche commented:
@aileronajay looking good! 
Some additional general remarks:

- Can you add a notice in the whatsnew file (doc/source/whatsnew/v0.20.0.txt)
- For the tests, you nicely structured them. But, if you would combine them in one test (you can keep the structure by using comments), you can shorten the code a large part by not having to repeat the expected values.
- Can you add docstrings to the new functions you defined? I know we don't always do that consistently in the codebase but we have to start somewhere (it doesn't always have to be long, eg a one line description can already do a lot)
- Not sure if this should be dealt with in this PR, but we should also make it possible to specify your own bins for these data types (now, you only tested for the case where you specify the number of bins, eg `bins=True`, while eg `bins=[pd.Timestamp(..), ...]` should also work)

aileronajay committed:
added docstring to new methods. Merged all datetime tests into one test, updated whatsnew

aileronajay commented:
@jorisvandenbossche I have addressed the review comments in https://github.com/pandas-dev/pandas/pull/14737/commits/13c49b24d4c01474c89edea9b36c7a09a26945cc

sinhrks commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -50,6 +50,7 @@ Other enhancements
 
 - ``pd.read_excel`` now preserves sheet order when using ``sheetname=None`` (:issue:`9930`)
 
+ENH: support cut/qcut for datetime/timedelta (GH14714)
```
pls follow the others style. like

> ``pd.cut`` and ``qcut`` now supports ``datetime64`` and ``timedelta64`` dtypes (:issue:`14714`)

aileronajay committed:
modified the whatsnew message

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -50,6 +50,7 @@ Other enhancements
 
 - ``pd.read_excel`` now preserves sheet order when using ``sheetname=None`` (:issue:`9930`)
 
+- pd.cut and qcut now support datetime64 and timedelta64 dtypes (issue:`14714`)
```
use double-backticks, e.g. ``pd.cut`` and say ``pd.qcut``

aileronajay replied:
made the changes now, just curious, is there a reason behind the double ticks? Does it get parsed by some code?

jorisvandenbossche replied:
@aileronajay yes, this is rst (restructuredText) used by sphinx to generate the html docs: http://www.sphinx-doc.org/en/1.4.9/rest.html

jreback commented on pandas/tools/tile.py:
```diff
@@ -300,3 +308,56 @@ def _trim_zeros(x):
+
+
+def _coerce_to_type(x):
+    '''
```
use triple-double quotes here, e.g. ``"""`` (our standard for doc-strings)

aileronajay replied:
@jreback changed it now

aileronajay committed:
changed docstring and whatsnew

aileronajay commented:
@jreback i have made the documentation changes requested

aileronajay committed:
removed trailing whitespace from test_tile file

aileronajay commented:
@jreback are there additional changes required in the PR?

jorisvandenbossche commented on pandas/tools/tile.py:
```diff
@@ -81,6 +82,11 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
+
     if not np.iterable(bins):
         if is_scalar(bins) and bins < 1:
             raise ValueError("`bins` should be a positive integer.")
```
The `try ... except ...` below should not be necessary anymore, since `x` is already an array (asarray is used in `_preprocess_for_cut`)

jorisvandenbossche replied:
yes (I couldn't comment on those lines) 
Normally it should already be assured before that you have an array here

aileronajay replied:
@jorisvandenbossche thanks, that is what i thought!

aileronajay replied:
I have addressed this review comment

jorisvandenbossche commented on pandas/tools/tile.py:
```diff
@@ -166,26 +175,26 @@ def qcut(x, q, labels=None, retbins=False, precision=3):
-            name = x.name
 
-    x = np.asarray(x)
+def _bins_to_cuts(x, bins, right=True, labels=None, retbins=False,
```
I think `retbins` arg is not used anymore in this function?

jorisvandenbossche replied:
And the same for `name`

jorisvandenbossche commented on pandas/tools/tile.py:
```diff
@@ -81,6 +82,11 @@ def cut(x, bins, right=True, labels=None, retbins=False, precision=3,
+    # for handling the cut for datetime and timedelta objects
+    x_is_series, series_index, name, x = _preprocess_for_cut(x)
+
+    original, x, dtype = _coerce_to_type(x)
```
I don't think `original` is used? In that case, it's not needed to return it

aileronajay replied:
I have addressed this review comment

jorisvandenbossche commented on pandas/tools/tile.py:
```diff
@@ -300,3 +308,56 @@ def _trim_zeros(x):
+    name = None
+    if x_is_series:
+        series_index = x.index
+        if name is None:
```
this check is not needed, as you defined `name = None` above

aileronajay replied:
I have addressed this review comment

jorisvandenbossche commented on pandas/tools/tile.py:
```diff
@@ -11,8 +11,9 @@
-
+from pandas import to_timedelta, to_datetime
 import numpy as np
+from pandas.types.common import is_datetime64_dtype, is_timedelta64_dtype
```
can you put this one above the numpy import? (to keep the pandas imports grouped like it was before)

aileronajay replied:
I have addressed this review comment

jorisvandenbossche reviewed (commented):
@aileronajay a few more small comments. After that it should be ready to merge I think! (we gave you already enough rounds of comments :-))

aileronajay committed:
code review comments

jorisvandenbossche committed:
some whitespace

jorisvandenbossche commented:
@aileronajay Thanks a lot! 
Do you want to do a follow-up PR to also add the ability to specify custom bins for datetime/timedelta values?

aileronajay commented:
@jorisvandenbossche thanks! Yes i would like to do PR. I will need some additional information though as to how the custom bins for datetime/timedelta values would work, i mean what would be the expected functionality

jorisvandenbossche commented:
@aileronajay The idea is the following. Assume the example from your tests:

```
In [2]: data = pd.to_datetime(pd.Series(['2013-01-01', '2013-01-02', '2013-01-03']))

In [6]: pd.cut(data, bins=3)
Out[6]: 
0    (2012-12-31 23:57:07.200000, 2013-01-01 16:00:00]
1           (2013-01-01 16:00:00, 2013-01-02 08:00:00]
2           (2013-01-02 08:00:00, 2013-01-03 00:00:00]
dtype: category
Categories (3, object): [(2012-12-31 23:57:07.200000, 2013-01-01 16:00:00] < (2013-01-01 16:00:00, 2013-01-02 08:00:00] <
                         (2013-01-02 08:00:00, 2013-01-03 00:00:00]]
```

Here you specify the number of bins, and the bin edges are generated automatically. But you can also specify the bin edges yourself (for numeric dtypes, this is not yet working for datetimes). So the idea is that you would be able to specify it like this:
```
In [7]: pd.cut(data, bins=[pd.Timestamp('2012-12-31'), pd.Timestamp('2012-01-02'), pd.Timestamp('2012-01-04')])
```

aileronajay commented:
@jorisvandenbossche, i think if we specify the bin (edges) in terms of datetime, the "data" or the input should also be of the type datetime, is that correct? Also we would only be able to specify edges only in terms of datetime and not timedelta, as datetime represent specific time instances and so between the datetime instances we have windows that we can split the input data into. Timedelta objects are difference in time objects, would be able to use them as bins?

jorisvandenbossche commented:
>  if we specify the bin (edges) in terms of datetime, the "data" or the input should also be of the type datetime, is that correct?

Yes, indeed, the specified bin edges need to match the type of the data

> we would only be able to specify edges only in terms of datetime and not timedelta

No, you added support for 'cutting' timedelta data in this PR:

```
In [6]: data = pd.Series(pd.timedelta_range("0 days", freq='H', periods=5))

In [7]: data
Out[7]: 
0   00:00:00
1   01:00:00
2   02:00:00
3   03:00:00
4   04:00:00
dtype: timedelta64[ns]

In [8]: pd.cut(data, bins=3)
Out[8]: 
0    (-1 days +23:59:45.600000, 0 days 01:20:00]
1    (-1 days +23:59:45.600000, 0 days 01:20:00]
2             (0 days 01:20:00, 0 days 02:40:00]
3             (0 days 02:40:00, 0 days 04:00:00]
4             (0 days 02:40:00, 0 days 04:00:00]
dtype: category
Categories (3, object): [(-1 days +23:59:45.600000, 0 days 01:20:00] < (0 days 01:20:00, 0 days 02:40:00] <
                         (0 days 02:40:00, 0 days 04:00:00]]
```

so it should also be possible to specify the bin edges manually for this type of data.

Actually, I see you didn't add tests for the timedelta case, and I didn't notice this before merging. Would you also like to do a PR for adding some tests for this (can be separate PR)?

aileronajay commented:
@jorisvandenbossche yes, i will do separate PR for tests for the timedelta case

aileronajay commented:
@jorisvandenbossche i coerced the bins to a numeric type, using the coerce private method that I created earlier and i get the following output (below) , is this on lines of what we desire?

 pd.cut(pd.Series([np.datetime64('2013-01-01'),np.datetime64('2013-01-02')]),
...     [np.datetime64('2012-12-18'), np.datetime64('2013-01-03'),np.datetime64('2013-01-05')])
0    (2012-12-18 00:00:00, 2013-01-03 00:00:00]
1    (2012-12-18 00:00:00, 2013-01-03 00:00:00]
dtype: category
Categories (2, object): [(2012-12-18 00:00:00, 2013-01-03 00:00:00] < (2013-01-03 00:00:00, 2013-01-05 00:00:00]]

jorisvandenbossche commented:
@aileronajay Yes, that looks like it. But you can just put up a PR, even if the code is only work in progress, then it will be easier to judge.

aileronajay commented:
@jorisvandenbossche , yeah about to do that now
